File: src/test/java/com/dozycoffee/wms/warehouse/fixture/LocationTestBuilder.java
package com.dozycoffee.wms.warehouse.fixture;

import com.dozycoffee.wms.warehouse.domain.enumeration.LocationStatus;
import com.dozycoffee.wms.warehouse.domain.model.Location;
import com.dozycoffee.wms.warehouse.domain.valueobject.Capacity;
import com.dozycoffee.wms.warehouse.domain.valueobject.LocationCode;

public class LocationTestBuilder {

    private Long locationId = null;
    private Long zoneId = 1L;
    private String locationCode = "A-01";
    private int maxCapacity = 70;
    private int usedCapacity = 0;
    private LocationStatus locationStatus = LocationStatus.ACTIVE;

    public static LocationTestBuilder location() {
        return new LocationTestBuilder();
    }

    public LocationTestBuilder locationId(Long locationId) {
        this.locationId = locationId;
        return this;
    }

    public LocationTestBuilder zoneId(Long zoneId) {
        this.zoneId = zoneId;
        return this;
    }

    public LocationTestBuilder locationCode(String locationCode) {
        this.locationCode = locationCode;
        return this;
    }

    public LocationTestBuilder maxCapacity(int maxCapacity) {
        this.maxCapacity = maxCapacity;
        return this;
    }

    public LocationTestBuilder usedCapacity(int usedCapacity) {
        this.usedCapacity = usedCapacity;
        return this;
    }

    public LocationTestBuilder locationStatus(LocationStatus locationStatus) {
        this.locationStatus = locationStatus;
        return this;
    }

    public Location build() {
        if (locationId != null) {
            return Location.reconstitute(
                    locationId,
                    zoneId,
                    new LocationCode(locationCode),
                    new Capacity(maxCapacity),
                    usedCapacity,
                    locationStatus
            );
        }
        return Location.create(
                zoneId,
                locationCode,
                maxCapacity,
                locationStatus
        );
    }
}

File: src/test/kotlin/com/dozycoffee/wms/warehouse/fixture/LocationTestBuilder.kt
package com.dozycoffee.wms.warehouse.fixture

import com.dozycoffee.wms.warehouse.domain.enumeration.AvailabilityStatus
import com.dozycoffee.wms.warehouse.domain.model.Location
import com.dozycoffee.wms.warehouse.domain.valueobject.Capacity
import com.dozycoffee.wms.warehouse.domain.valueobject.LocationCode

class LocationTestBuilder {

    private var locationId: Long? = null
    private var zoneId: Long? = 1L
    private var locationCode: String? = "A-01"
    private var maxCapacity: Int = 70
    private var usedCapacity: Int = 0
    private var locationStatus: AvailabilityStatus? = AvailabilityStatus.AVAILABLE

    companion object {
        fun location(): LocationTestBuilder = LocationTestBuilder()
    }

    fun locationId(locationId: Long?): LocationTestBuilder {
        this.locationId = locationId
        return this
    }

    fun zoneId(zoneId: Long?): LocationTestBuilder {
        this.zoneId = zoneId
        return this
    }

    fun locationCode(locationCode: String?): LocationTestBuilder {
        this.locationCode = locationCode
        return this
    }

    fun maxCapacity(maxCapacity: Int): LocationTestBuilder {
        this.maxCapacity = maxCapacity
        return this
    }

    fun usedCapacity(usedCapacity: Int): LocationTestBuilder {
        this.usedCapacity = usedCapacity
        return this
    }

    fun locationStatus(locationStatus: AvailabilityStatus?): LocationTestBuilder {
        this.locationStatus = locationStatus
        return this
    }

    fun build(): Location {
        val id = locationId
        if (id != null) {
            return Location.reconstitute(
                id,
                requireNotNull(zoneId),
                LocationCode.of(locationCode),
                Capacity(maxCapacity),
                usedCapacity,
                requireNotNull(locationStatus)
            )
        }
        return Location.create(zoneId, locationCode, maxCapacity, locationStatus)
    }
}
